### PR TITLE
macros: improve the `TokenStream` quoter

### LINKED ISSUE #39746: token quote should be able to unquote a single token
At the moment it can only unquote a `TokenStream`.

E.g., this should work:

```
let f = token::Ident("foo");
qquote!(let unquote f = blah;);
```

Also, we should have syntax for unquote, but that is unrelated.

### PR DESCRIPTION
**Pull request by jseyfried**
This PR
 - renames the `TokenStream` quoter from `qquote!` to `quote!`,
 - uses `$` instead of `unquote` (e.g. `let toks: TokenStream = ...; quote!([$toks])`),
 - allows unquoting `Token`s as well as `TokenTree`s and `TokenStream`s (fixes #39746), and
 - to preserve syntactic space, requires that `$` be followed by
   - a single identifier to unquote, or
   - another `$` to produce a literal `$`.

r? @nrc

### CODE REVIEW

File: src/test/run-pass-fulldeps/auxiliary/hello_macro.rs [hunk 1]
```diff
@@ -29,6 +29,11 @@ pub fn plugin_registrar(reg: &mut Registry) {
 
 // This macro is not very interesting, but it does contain delimited tokens with
 // no content - `()` and `{}` - which has caused problems in the past.
+// Also, it tests that we can escape `$` via `$$`.
```
Comment (nrc) [inline src/test/run-pass-fulldeps/auxiliary/hello_macro.rs]: Could you document this in the doc comments for the quote module please?

**Review by nrc — APPROVED** [review]
Is there any clash between this quote! macro and the AST quote! macro? I don't recall if that is magically in the global namespace or anything.

r=me if not and with the $$ doc

File: src/test/run-pass-fulldeps/auxiliary/hello_macro.rs (hunk 1 above)
Comment (jseyfried) [inline src/test/run-pass-fulldeps/auxiliary/hello_macro.rs]: Will do.

**Comment by jseyfried** [comment]
@nrc None of the [AST quoters](https://github.com/rust-lang/rust/blob/212b6c25507b963b60a646a2ff3df7496bd30acf/src/libsyntax_ext/lib.rs#L72-L88) are named `quote`.

**Pull request merged**